### PR TITLE
#48475 Pure disjunctions should rewrite to a MatchNoneQueryBuilder

### PR DESCRIPTION
**Pull request by SivagurunathanV**
1. Added Pure disjunctions should rewrite to a MatchNoneQueryBuilder when all instance matches MatchNoneQueryBuilder
2. Added testcase for the same.

### CODE REVIEW

**Comment by SivagurunathanV** [comment]
#48557

**Comment by jimczi** [comment]
@SivagurunathanV we already [rewrite](https://github.com/elastic/elasticsearch/pull/48557/files#diff-b627565803e5742638a07a2c45fde79dR424-R425) `bool` query with a `MatchNoneQueryBuilder` in a `filter` or a `must` clause to a match none  query so I don't think that your pr changes anything. We don't need to check that all required clauses are an instance of `MatchNoneQueryBuilder`, one is enough.

**Comment by jpountz** [comment]
@SivagurunathanV The linked issue is about optional clauses while your PR handles required clauses. If you don't understand how the issue should be addressed, please raise questions before opening a pull request.

**Comment by SivagurunathanV** [comment]
@jimczi Thanks for the clarification. @jpountz yeah sure. Can you please confirm, `optional` clauses you mean the `should` BoolQuery classes alone?

**Comment by SivagurunathanV** [comment]
@jpountz Thanks for the explanation. Based on my understanding, if the required clauses are empty and the optional clause is returning MatchNone, we can safely make an early termination.

**Comment by jpountz** [comment]
@SivagurunathanV Yes, if there are no required clauses and _all_ SHOULD clauses rewrite to a MatchNoneQueryBuilder then the bool query can rewrite to a MatchNoneQueryBuilder.

**Comment by SivagurunathanV** [comment]
@jpountz Thanks. Added my changes to the PR. Please let me know if that works.

**Review by jpountz — APPROVED** [review]
Thanks, the change looks good to me! I'll merge this PR next week.

**Comment by SivagurunathanV** [comment]
@jpountz Can you merge this PR?

**Comment by jpountz** [comment]
Oops, thanks for the reminder, I'll merge soon.

**Comment by jpountz** [comment]
@elasticmachine test it please

**Comment by jpountz** [comment]
@elasticmachine ok to test

File: server/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java [hunk 1]
```diff
@@ -417,12 +417,19 @@ protected QueryBuilder doRewrite(QueryRewriteContext queryRewriteContext) throws
         changed |= rewriteClauses(queryRewriteContext, mustNotClauses, newBuilder::mustNot);
         changed |= rewriteClauses(queryRewriteContext, filterClauses, newBuilder::filter);
         changed |= rewriteClauses(queryRewriteContext, shouldClauses, newBuilder::should);
+        // early termination when must clause is empty and optional clauses is returning MatchNoneQueryBuilder
+        if(mustClauses.size() == 0 && filterClauses.size() == 0
```
Comment (jpountz) [inline server/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java]: Given test failures, I think the problem is that this rewrites empty boolean queries to a match_none, while this is currently treated as a match_all. We probably to do something like below, can you add a unit test for this case?

```suggestion
        if (mustClauses.size() == 0 && filterClauses.size() == 0 && shouldClauses.size() > 0
```

Comment (SivagurunathanV) [reply server/src/main/java/org/elasticsearch/index/query/BoolQueryBuilder.java]: Thanks for the catch. I just updated the code. Let me know.

**Comment by SivagurunathanV** [comment]
@jpountz I am not able to see the failed job logs in Jenkins, can you please help me here. what is the exact issue. Thanks

**Comment by jpountz** [comment]
@elasticmachine ok to test

**Comment by jpountz** [comment]
@SivagurunathanV  The logs don't require authentication or anything, so you should be able to read them. Some tests were complaining that the matching documents were not the expected ones.

**Comment by jpountz** [comment]
@elasticmachine update master

**Comment by jpountz** [comment]
@elasticmachine update branch

**Comment by jpountz** [comment]
@SivagurunathanV Would you mind merging master into your branch?

**Comment by SivagurunathanV** [comment]
@jpountz Thanks for the help. I merged the master into my branch and all test passed :)

**Comment by jpountz** [comment]
@elasticmachine ok to test

**Comment by jpountz** [comment]
Thanks @SivagurunathanV !

**Pull request merged**